pkg/audit/audit.go

```diff
@@ -0,0 +1,139 @@
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	k8stypes "k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	contentTypeJSON = "application/json"
+)
+
+const (
+	levelNull = iota
+	levelMetadata
+	levelRequest
+	levelRequestResponse
+)
+
+var (
+	bodyMethods = map[string]bool{
+		http.MethodPut:  true,
+		http.MethodPost: true,
+	}
+)
+
+type auditLog struct {
+	log     *log
+	writer  *LogWriter
+	reqBody []byte
+}
+
+type log struct {
+	AuditID           k8stypes.UID `json:"auditID,omitempty"`
+	RequestURI        string       `json:"requestURI,omitempty"`
+	User              *userInfo    `json:"user,omitempty"`
+	Verb              string       `json:"verb,omitempty"`
+	SourceIP          string       `json:"sourceIP,omitempty"`
+	RequestTimestamp  string       `json:"requestTimestamp,omitempty"`
+	ResponseTimestamp string       `json:"responseTimestamp,omitempty"`
+	ResponseCode      int          `json:"responseCode,omitempty"`
+	RequestHeader     http.Header  `json:"requestHeader,omitempty"`
+	ResponseHeader    http.Header  `json:"responseHeader,omitempty"`
+	RequestBody       []byte       `json:"requestBody,omitempty"`
+	ResponseBody      []byte       `json:"responseBody,omitempty"`
+}
+
+type userInfo struct {
+	Name  string `json:"name,omitempty"`
+	Group string `json:"group,omitempty"`
+}
+
+func new(writer *LogWriter, req *http.Request) (*auditLog, error) {
+	auditLog := &auditLog{
+		writer: writer,
+		log: &log{
+			AuditID:          k8stypes.UID(uuid.NewRandom().String()),
+			RequestURI:       req.RequestURI,
+			Verb:             req.Method,
+			SourceIP:         getSourceIPs(req.RemoteAddr),
+			RequestTimestamp: time.Now().Format(time.RFC3339),
+		},
+	}
+
+	contentType := req.Header.Get("Content-Type")
+	if writer.Level >= levelRequest && bodyMethods[req.Method] && contentType == contentTypeJSON {
```

Comment:
Do you really care if it's JSON?

Reply:
I afraid other format logs will make the audit log mess, hard to parse in other tools.